celery/app/amqp.py

```diff
@@ -521,7 +521,7 @@ def send_task_message(producer, name, message,
                     exchange_type = 'direct'
 
             # convert to anon-exchange, when exchange not set and direct ex.
-            if not exchange or not routing_key and exchange_type == 'direct':
+            if (not exchange or not routing_key) and exchange_type == 'direct':
```

Comment:
I see a bunch of `test_send_task_message__*` methods in [https://github.com/celery/celery/blob/522ef610ea17ae79695e995ddd82d034e8e8bc7a/t/unit/app/test_amqp.py](t/unit/app/test_amqp.py), so I'll start there...